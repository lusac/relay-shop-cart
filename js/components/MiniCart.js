import React from 'react'

export default class MiniCart extends React.Component {
  handleToggle() {
    if (!this.props.show)
      return 'hidden'
  }

  render() {
    return (
      <div className={'mini-cart ' + this.handleToggle()}>
        <div className="mini-cart__header">carrinho</div>
        <div className="mini-cart__section">
          <ul className="mini-cart__items">
            {Object.keys(this.props.items).map((id) => {
              return (
                <li key={id} className="mini-cart__item">
                  <img
                    className="mini-cart__item__img"
                    src={'/imgs/' + this.props.items[id].image}></img>
                  <span className="mini-cart__item__middle">
                    <h5 className="mini-cart__item__name">
                      {this.props.items[id].name}
                    </h5>
                    <small className="mini-cart__item__qty">
                      Quantidade: {this.props.items[id].qty}
                    </small>
                  </span>
                  <strong className="mini-cart__item__total">
                    R$ {this.props.items[id].qty * this.props.items[id].price}
                  </strong>
                </li>
              )
            })}
          </ul>
        </div>
        <div className="mini-cart__footer">
          total R$ {this.props.totalPrice}
        </div>
      </div>
    )
  }
}