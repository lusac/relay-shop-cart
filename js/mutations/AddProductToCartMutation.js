import {commitMutation, graphql} from 'react-relay'

const mutation = graphql`
  mutation AddProductToCartMutation($input: AddProductToCartInput!) {
    addProductToCart(input: $input) {
      product {
        amount
      }
      viewer {
        id
        cart(
          first: 2147483647 # max GraphQLInt
        ) @connection(key: "Header_cart") {
          edges {
            node {
              id
              ...Product_product
            }
          }
        }
      }
    }
  }
`

function commit(environment, product, user) {
  return commitMutation(environment, {
    mutation,
    variables: {
      input: {id: product.id}
    }
  })
}

export default {commit}
