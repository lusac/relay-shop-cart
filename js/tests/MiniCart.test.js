import React from 'react'
import renderer from 'react-test-renderer'
import MiniCart from '../components/MiniCart'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import AddProductToCartMutation from '../mutations/AddProductToCartMutation'
import DecreaseProductToCartMutation from '../mutations/DecreaseProductToCartMutation'
import RemoveProductToCartMutation from '../mutations/RemoveProductToCartMutation'

configure({ adapter: new Adapter() })

describe('MiniCart component', () => {
  describe('snapshot', () => {
    it('MiniCart default', () => {
      const
        items=[
          {id: 1, name: 'item a', image: 'a.png', price: 11, amount: 1},
          {id: 2, name: 'item b', image: 'b.png', price: 22, amount: 2},
          {id: 3, name: 'item c', image: 'c.png', price: 33, amount: 3}
        ],
        renderedValue = renderer.create(<MiniCart items={items} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    let product
    let environment
    let items

    beforeEach(() => {
      product={
        id: 1, name: 'item a', image: 'a.png', price: 11, amount: 1, qty: 2
      },
      environment = {xunda:1},
      items=[
        {id: 1, name: 'item a', image: 'a.png', price: 11, amount: 1},
        {id: 2, name: 'item b', image: 'b.png', price: 22, amount: 2},
        {id: 3, name: 'item c', image: 'c.png', price: 33, amount: 3}
      ]
    })

    describe('method _handleIncreaseProduct', () => {
      it('Add product when amount bigger then 0', () => {
        let wrapper = shallow(<MiniCart items={items} environment={environment} />)
        spyOn(AddProductToCartMutation, 'commit')
        wrapper.instance()._handleIncreaseProduct(product)
        expect(AddProductToCartMutation.commit).toHaveBeenCalledWith(environment, product)
      })

      // Mockar Alert
      // it('Do not Add product when amount equal to 0', () => {
      //   const product = {
      //     id: 1, name: 'item a', image: 'a.png', price: 11, amount: 0
      //   }
      //   let wrapper = shallow(<MiniCart items={items} environment={environment} />)
      //   spyOn(AddProductToCartMutation, 'commit')
      //   wrapper.instance()._handleIncreaseProduct(product)
      //   expect(AddProductToCartMutation.commit).not.toHaveBeenCalled()
      // })
    })

    describe('method _handleDecreaseProduct', () => {
      it('Decrease product while qty bigger then 0', () => {
        let wrapper = shallow(<MiniCart items={items} environment={environment} />)
        spyOn(DecreaseProductToCartMutation, 'commit')
        wrapper.instance()._handleDecreaseProduct(product)
        expect(DecreaseProductToCartMutation.commit).toHaveBeenCalledWith(environment, product)
      })

      it('Do not decrease product while qty equal to 0', () => {
        product.qty = 0
        let wrapper = shallow(<MiniCart items={items} environment={environment} />)
        spyOn(DecreaseProductToCartMutation, 'commit')
        wrapper.instance()._handleDecreaseProduct(product)
        expect(DecreaseProductToCartMutation.commit).not.toHaveBeenCalled()
      })
    })

    describe('method _handleRemoveProduct', () => {
      it('Decrease product while qty bigger then 0', () => {
        let wrapper = shallow(<MiniCart items={items} environment={environment} />)
        spyOn(RemoveProductToCartMutation, 'commit')
        wrapper.instance()._handleRemoveProduct(product)
        expect(RemoveProductToCartMutation.commit).toHaveBeenCalledWith(environment, product)
      })
    })

    describe('method _handleToggle', () => {
      it('return class if show props is false', () => {
        let wrapper = shallow(<MiniCart show={false} items={items} environment={environment} />)
        let resp = wrapper.instance()._handleToggle()
        expect(resp).toBe('mini-cart--hidden')
      })

      it('return undefined if show props is true', () => {
        let wrapper = shallow(<MiniCart show={true} items={items} environment={environment} />)
        let resp = wrapper.instance()._handleToggle()
        expect(resp).toBeUndefined()
      })
    })
  })
})