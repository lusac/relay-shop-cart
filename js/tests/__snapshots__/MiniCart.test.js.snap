// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniCart component snapshot MiniCart default 1`] = `
<div
  className="mini-cart--hidden"
>
  <div
    className="mini-cart__overlay"
  />
  <div
    className="mini-cart"
  >
    <div
      className="mini-cart__header"
    >
      carrinho
    </div>
    <div
      className="mini-cart__section"
    >
      <ul
        className="mini-cart__items"
      >
        <li
          className="mini-cart__item"
        >
          <span
            className="mini-cart__item__remove"
            onClick={[Function]}
          >
            X
          </span>
          <img
            className="mini-cart__item__img"
            src="/imgs/a.png"
          />
          <span
            className="mini-cart__item__middle"
          >
            <h5
              className="mini-cart__item__name"
            >
              item a
            </h5>
            <small
              className="mini-cart__item__qty"
            >
              Quantidade:
              <span
                className="mini-cart__change-qty"
              >
                <span
                  onClick={[Function]}
                >
                  -
                </span>
                <input
                  type="text"
                />
                <span
                  onClick={[Function]}
                >
                  +
                </span>
              </span>
            </small>
          </span>
          <strong
            className="mini-cart__item__total"
          >
            R$ 
            NaN
          </strong>
        </li>
        <li
          className="mini-cart__item"
        >
          <span
            className="mini-cart__item__remove"
            onClick={[Function]}
          >
            X
          </span>
          <img
            className="mini-cart__item__img"
            src="/imgs/b.png"
          />
          <span
            className="mini-cart__item__middle"
          >
            <h5
              className="mini-cart__item__name"
            >
              item b
            </h5>
            <small
              className="mini-cart__item__qty"
            >
              Quantidade:
              <span
                className="mini-cart__change-qty"
              >
                <span
                  onClick={[Function]}
                >
                  -
                </span>
                <input
                  type="text"
                />
                <span
                  onClick={[Function]}
                >
                  +
                </span>
              </span>
            </small>
          </span>
          <strong
            className="mini-cart__item__total"
          >
            R$ 
            NaN
          </strong>
        </li>
        <li
          className="mini-cart__item"
        >
          <span
            className="mini-cart__item__remove"
            onClick={[Function]}
          >
            X
          </span>
          <img
            className="mini-cart__item__img"
            src="/imgs/c.png"
          />
          <span
            className="mini-cart__item__middle"
          >
            <h5
              className="mini-cart__item__name"
            >
              item c
            </h5>
            <small
              className="mini-cart__item__qty"
            >
              Quantidade:
              <span
                className="mini-cart__change-qty"
              >
                <span
                  onClick={[Function]}
                >
                  -
                </span>
                <input
                  type="text"
                />
                <span
                  onClick={[Function]}
                >
                  +
                </span>
              </span>
            </small>
          </span>
          <strong
            className="mini-cart__item__total"
          >
            R$ 
            NaN
          </strong>
        </li>
      </ul>
    </div>
    <div
      className="mini-cart__footer"
    >
      total R$ 
    </div>
  </div>
</div>
`;
